sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/service/ExecutionContext.java

```diff
@@ -0,0 +1,126 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.stargate.sgv2.docsapi.service;
+
+import io.stargate.bridge.proto.QueryOuterClass.Query;
+import io.stargate.sgv2.docsapi.models.ExecutionProfile;
+import io.stargate.sgv2.docsapi.models.ImmutableExecutionProfile;
+import io.stargate.sgv2.docsapi.models.QueryInfo;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.stream.Collectors;
+
+public abstract class ExecutionContext {
+
+  public static final ExecutionContext NOOP_CONTEXT = new NoOpContext();
+
+  public static ExecutionContext create(Boolean enabled) {
+    if (Boolean.TRUE.equals(enabled)) {
+      return new ProfilingContext("root");
+    } else {
+      return NOOP_CONTEXT;
+    }
+  }
+
+  public abstract ExecutionContext nested(String description);
+
+  /** Records information about a query when its results are retrieved. */
+  public abstract void traceCqlResult(Query query, int numRows);
+
+  /**
+   * Immediately records information about a DML query that is scheduled to be executed at a later
+   * time.
+   */
+  public abstract void traceDeferredDml(Query query);
+
+  public abstract ExecutionProfile toProfile();
+
+  private static class NoOpContext extends ExecutionContext {
+    @Override
+    public void traceCqlResult(Query query, int numRows) {
+      // nop
+    }
+
+    @Override
+    public void traceDeferredDml(Query query) {
+      // nop
+    }
+
+    @Override
+    public ExecutionContext nested(String description) {
+      return this;
+    }
+
+    @Override
+    public ExecutionProfile toProfile() {
+      return null;
+    }
+  }
+
+  private static class ProfilingContext extends ExecutionContext {
+    private final Queue<ProfilingContext> steps = new ConcurrentLinkedQueue<>();
+    private final Map<String, QueryInfo> executionInfoMap = new ConcurrentHashMap<>();
+    private final String description;
+
+    private ProfilingContext(String description) {
+      this.description = description;
+    }
+
+    @Override
+    public final ExecutionContext nested(String description) {
+      ProfilingContext step = new ProfilingContext(description);
+      steps.add(step);
+      return step;
+    }
+
+    @Override
+    public void traceCqlResult(Query query, int numRows) {
+      String cql = query.getCql();
+      executionInfoMap.merge(cql, QueryInfo.of(cql, numRows), QueryInfo::combine);
+    }
+
+    @Override
+    public void traceDeferredDml(Query query) {
+      int numRows = 0;
+
+      // TODO the query is already built, so we have no way of estimating the rows impacted
```

Comment:
I'm not sure how we can implement this in the v2 architecture.

Reply:
So I guess I'll just remove `QueryInfo.rowCount()`.

Reply:
The v1 implementation does not return an exact count either:
```java
RowsImpacted rowsImpacted = ((BoundDMLQuery) query).rowsUpdated();
if (rowsImpacted.isKeys()) {
  numRows = rowsImpacted.asKeys().primaryKeys().size();
}
```

For inserts, this works. But also, the count is always 1.
For deletes:
- `DeleteSubDocumentKeysQueryBuilder` uses an IN clause, which only returns an exact count _if that clause covers the last clustering column_. In our model this is unlikely to happen: a subdocument based at the last _pX_could only be empty. So in practice the IN condition will be on a higher-level column, the delete will be considered a range query, and therefore we'll count 0 rows (`isKeys()) == false`).
- other deletes are always range queries, so they return 0 too.

Bottom line: we can hard-code `numRows` to 1 and be at least as accurate as v1.


<details>
<summary>v1 test code</summary>

```java
package io.stargate.db.query.builder;

import com.google.common.collect.ImmutableList;
import io.stargate.db.query.BoundDMLQuery;
import io.stargate.db.query.Predicate;
import io.stargate.db.query.RowsImpacted;
import io.stargate.db.query.TypedValue;
import io.stargate.db.schema.Column;
import io.stargate.db.schema.Schema;

public class Tmp {

  private static final Schema SCHEMA =
      Schema.build()
          .keyspace("ks")
          .table("tbl")
          .column("id", Column.Type.Text, Column.Kind.PartitionKey)
          .column("p0", Column.Type.Text, Column.Kind.Clustering)
          .column("p1", Column.Type.Text, Column.Kind.Clustering)
          .column("p2", Column.Type.Text, Column.Kind.Clustering)
          .column("v", Column.Type.Text, Column.Kind.Regular)
          .build();

  public static void main(String[] args) {
    QueryBuilder queryBuilder = new QueryBuilder(SCHEMA, TypedValue.Codec.testCodec(), null);
    BoundDMLQuery bound =
        ((BoundDMLQuery)
            queryBuilder
                .delete()
                .from("ks", "tbl")
                .where("id", Predicate.EQ)
                .where("p0", Predicate.IN)
                .build()
                .bind("id", ImmutableList.of("a", "b", "c")));
    RowsImpacted rowsImpacted = bound.rowsUpdated();

    System.out.printf(
        "IsKeys: %s, estimated count: %d%n",
        rowsImpacted.isKeys(),
        rowsImpacted.isKeys() ? rowsImpacted.asKeys().primaryKeys().size() : 0);
  }
}
```

</details>

Reply:
Good  investigation.. Let's do it then like that.. Can we just update the API documentation in the `QueryInfo`:

```
 @Schema(description = "The total number of rows fetched by this query")
  @JsonProperty("rowCount")
  int rowCount();
```

So mention this is approximate number of rows impacted or smth like that